msg.go

```diff
@@ -923,58 +923,55 @@ func (dns *Msg) String() string {
 // than packing it, measuring the size and discarding the buffer.
 func (dns *Msg) Len() int { return compressedLen(dns, dns.Compress) }
 
-func compressedLenWithCompressionMap(dns *Msg, compression map[string]int) int {
-	l := 12 // Message header is always 12 bytes
-	for _, r := range dns.Question {
-		compressionLenHelper(compression, r.Name, l)
-		l += r.len()
-	}
-	l += compressionLenSlice(l, compression, dns.Answer)
-	l += compressionLenSlice(l, compression, dns.Ns)
-	l += compressionLenSlice(l, compression, dns.Extra)
-	return l
-}
-
 // compressedLen returns the message length when in compressed wire format
 // when compress is true, otherwise the uncompressed length is returned.
 func compressedLen(dns *Msg, compress bool) int {
-	// We always return one more than needed.
+	lenFn := Len
 	if compress {
-		compression := map[string]int{}
-		return compressedLenWithCompressionMap(dns, compression)
+		lenFn = newCompressionLenFn(dns.Question)
 	}
-	l := 12 // Message header is always 12 bytes
 
-	for _, r := range dns.Question {
-		l += r.len()
+	l := 12 // Message header is always 12 bytes
+	for _, q := range dns.Question {
+		l += q.len()
 	}
 	for _, r := range dns.Answer {
 		if r != nil {
-			l += r.len()
+			l += lenFn(r)
 		}
 	}
 	for _, r := range dns.Ns {
 		if r != nil {
-			l += r.len()
+			l += lenFn(r)
 		}
 	}
 	for _, r := range dns.Extra {
 		if r != nil {
-			l += r.len()
+			l += lenFn(r)
 		}
 	}
 
 	return l
 }
 
-func compressionLenSlice(lenp int, c map[string]int, rs []RR) int {
-	initLen := lenp
-	for _, r := range rs {
+// newCompressionLenFn returns a function which returns the compressed length of
+// a given resource record.
+//
+// Calling the returned length function is not side-effect free, as the name of
+// the record might be added to the compression dictionary, and should only be
+// used for records of the same message.
+func newCompressionLenFn(qs []Question) func(RR) int {
```

Comment:
This will only carry 1 question at most, and looking at the code we only use the `q.Name`, so I think we can just use a string here, and drop the range below